trace/config.go

```diff
@@ -203,3 +205,56 @@ func (i instrumentationVersionOption) ApplyTracer(config *TracerConfig) {
 }
 
 func (instrumentationVersionOption) private() {}
+
+// ErrorConfig is a group of options for error.
+type ErrorConfig struct {
+	Code    codes.Code
+	Message string
+	// EventOpts will be passed to addEvent
+	EventOpts []EventOption
+}
+
+// ErrorOption applies an option to a ErrorConfig.
+type ErrorOption interface {
+	ApplyError(config *ErrorConfig)
+
+	// A private method to prevent users implementing the
+	// interface and so future additions to it will not
+	// violate compatibility.
+	private()
+}
+
+// NewErrorConfig applies all the options to a returned ErrorConfig.
+func NewErrorConfig(options ...ErrorOption) *ErrorConfig {
+	c := new(ErrorConfig)
+	for _, option := range options {
+		option.ApplyError(c)
+	}
```

Comment:
I'd like to avoid the need to use `WithEventOpts()` when calling `RecordError()`.  My first thought was that maybe something like this would work: 

 ```go
	for _, option := range options {
		if evopt, ok := option.(EventOption); ok {
			c.EventOpts = append(c.EventOpts, option)
		} else {
			option.ApplyError(c)
		}
	}
```

But I don't think it does since the method signature now requires `ErrorOption`s to be passed instead of `EventOption`s.  So, my next thought is could the relevant `EventOption` implementations gain an `ErrorOption` implementation that looks something like this:

```go
func (o someOptionType) ApplyError(c *ErrorConfig) {
	c.EventOpts = append(c.EventOpts, o)
}
```

Reply:
I don't like the wrap either,  I had tried this  way to  make it work before the wrap.

But it can't work, beacuse the compile complains about the type mismatch.
Error message:
     Cannot use 'trace.WithTimestamp(errTime)' (type LifeCycleOption) as the type ErrorOption Type does not implement 
   'ErrorOption' as some methods are missing: ApplyError(config *ErrorConfig)

Code looks like this:

```
config:
 type timestampSpanOption time.Time

func (o timestampSpanOption) ApplySpan(c *SpanConfig)  { o.apply(c) }
func (o timestampSpanOption) ApplyEvent(c *SpanConfig) { o.apply(c) }
func (o timestampSpanOption) ApplyError(c *ErrorConfig) {c.Timestamp = time.Time(o)}
func (timestampSpanOption) private()                   {}
func (o timestampSpanOption) apply(c *SpanConfig)      { c.Timestamp = time.Time(o) }



client code:
span.RecordError(s.err, trace.WithTimestamp(errTime))


RecordError:

func (s *span) RecordError(err error, options ...trace.ErrorOption) {
	if s == nil || err == nil || !s.IsRecording() {
		return
	}
	c := trace.NewErrorConfig(options...)
	if c.Code != codes.Unset {
		s.SetStatus(c.Code, c.Message)
	}

	var eventOpts []trace.EventOption
	for _, option := range options {
		if option, ok := option.(trace.EventOption); ok {
			eventOpts = append(eventOpts, option)
		}
	}
	eventOpts = append(eventOpts, trace.WithAttributes(
		semconv.ExceptionTypeKey.String(typeStr(err)),
		semconv.ExceptionMessageKey.String(err.Error()),
	))
	s.addEvent(semconv.ExceptionEventName, eventOpts...)
}

```



I don't think it's a good idea to make LifeCycleOption implements ErrorOption.

If you know any idea, that will be a great help, I'd like to remove the wrap.

Reply:
Why not just embed the `ErrorOption` in the `EventOption`? https://github.com/open-telemetry/opentelemetry-go/blob/main/CONTRIBUTING.md#dealing-with-overlap